### DISCUSSION TITLE
Why multiple actions are allowed for non-root actions but only one action with root access?

### DISCUSSION THREAD

**Discussion by akshay-ap** [5afe/safe-core-protocol-specs#6] [Q&A]
Mediator spec: https://github.com/5afe/safe-protocol-specs/tree/main/mediator#interface

**Comment by rmeissner** [comment]
One of the most used functionality of delegate calls with the Safe contracts is to batch transactions. Therefore providing this functionality to all modules that interact with the protocol is a given.

On the other side root access is a critical interaction, being able to nest this make it harder to understand what is going on. If this functionality (i.e. nested delegate calls) is required it is possible to "load" this via root access call to a library (i.e. MultiSend). 

So to minimize the complexity in the root access path the specs proposed to only allow one "root access action" at a time.

**Reply by auryn-macmillan** [reply]
For the sake of interface consistency, and so that folks aren't forced to use an external library for batch Root actions, I'd very much prefer SafeRootAccess() to also be a batch action.

> root access is a critical interaction, being able to nest this make it harder to understand what is going on

I actually disagree on this. If users must call a library to do batches of root actions, then this adds a layer of encoding to the transactions, making them less human readable.

**Comment by auryn-macmillan** [comment]
@rmeissner, let's say I want to create a batch like:
1. approve Dai
2. delegate call to the [CowswapOrderSigner](https://github.com/gnosis/cow-order-signer/blob/master/contracts/CowswapOrderSigner.sol)

How would you envision constructing this batch with the `executeTransaction()` and `executeRootAccess()` entrypoints? Does this imply using `executeRootAccess() and an external multisend library?

**Reply by mmv08** [reply]
The order signer doesn't require a delegate call, does it? 
https://github.com/gnosis/cow-order-signer/blob/master/contracts/cowProtocol/mixins/GPv2Signing.sol#L90

**Reply by mmv08** [reply]
The non-root flow supports batching our of the box

**Reply by auryn-macmillan** [reply]
`signOrder()` [requires a delegate call](https://github.com/gnosis/cow-order-signer/blob/f27b30d427c7e51963849e06e1edffa352a21e22/contracts/CowswapOrderSigner.sol#L77), this function is designed to allow us to set permissions via the [roles mod](https://github.com/gnosis/zodiac-modifier-roles) where you can scope the parameters of the order; to do things like enforce who the recipient of the order is.
